File: 백준/Platinum/1725. 히스토그램/히스토그램.cc
#pragma warning(disable:4996)
#include<stdio.h>
#include<algorithm>
#define MAX(x, y)( ((x)>(y)) ? (x) : (y) )
#define M 0x7fffffff
using namespace std;
typedef pair<int, int> temp;
int n, cnt, a, i, ans;
temp seg[1 << 18];
temp MIN(temp a, temp b) { return a.first < b.first ? a : b; }
temp insert(int left, int right, int idx)
{
	int mid = (left + right) / 2;
	if (i < left || right < i) return seg[idx];
	if (left == i && right == i) return seg[idx] = {a, i};
	return seg[idx] = MIN(insert(left, mid, idx * 2), insert(mid + 1, right, idx * 2 + 1));
}
temp query(int left, int right, int idx, int st, int ed)
{
	int mid = (left + right) / 2;
	if (ed < left || right < st) return make_pair(M, M);
	if (st <= left && right <= ed) return seg[idx];
	return MIN(query(left, mid, idx * 2, st, ed), query(mid + 1, right, idx * 2 + 1, st, ed));
}
void 분할(int left, int right)
{
	if (left > right) return;
	temp tmp = query(1, cnt, 1, left, right);
	ans = MAX(ans, tmp.first * (right - left + 1));
	분할(left, tmp.second - 1);
	분할(tmp.second + 1, right);
}
int main()
{
	scanf("%d", &n);
	for (cnt = 1; cnt <= n; cnt *= 2);
	for (i = 1; i <= n; i++)
	{
		scanf("%d", &a);
		insert(1, cnt, 1);
	}
	분할(1, n);
	printf("%d", ans);
	return 0;
}

File: 백준/Platinum V/1725. 히스토그램/히스토그램.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(60000)
def MIN(a, b):
    return a if a[1] < b[1] else b
def insert(index, value):
    global idx
    temp = idx + index - 1
    seg[temp] = [index, value]
    temp //= 2
    while temp > 0:
        seg[temp] = MIN(seg[temp * 2], seg[temp * 2 + 1])
        temp //= 2
def query(left, right, idx, st, ed):
    global MAX
    if ed < left or right < st : return [MAX, MAX]
    if st <= left and right <= ed : return seg[idx]
    mid = (left + right) // 2
    return MIN(query(left, mid, idx * 2, st, ed), query(mid + 1, right, idx * 2 + 1, st, ed))

def 분할(left, right):
    global n, ans, idx
    if left > right : return
    temp = query(1, idx, 1, left, right)
    ans = max(ans, temp[1] * (right - left + 1))

    분할(left, temp[0] - 1)
    분할(temp[0] + 1, right)
n = int(input())
MAX = 0x7fffffff
seg = [[0, MAX] for _ in range(1 << 18)]

idx = 1
ans = 0
while idx <= n :
    idx *= 2
for i in range(1, n + 1):
    m = int(input())
    insert(i, m)
check = 1
cnt = 1
분할(1, n)
print(ans)